File: 프로그래머스/3/42895. N으로 표현/N으로 표현.cpp
#include <iostream>
#include <vector>
#include <set>

using namespace std;

int solution(int N, int number) {
    if (N == number) return 1;  // N을 1번 사용한 경우
    
    vector<set<int>> dp(9);  // dp[i]: N을 i번 사용해서 만들 수 있는 숫자 집합

    for (int i = 1; i <= 8; i++) {
        // N을 i번 연속해서 쓴 값 추가 (예: 5, 55, 555, 5555, ...)
        int num = 0;
        for (int j = 0; j < i; j++) {
            num = num * 10 + N;
        }
        dp[i].insert(num);

        // i를 두 부분 (j, i-j)로 나눠서 사칙연산 수행
        for (int j = 1; j < i; j++) {
            for (int a : dp[j]) {
                for (int b : dp[i - j]) {
                    dp[i].insert(a + b);
                    dp[i].insert(a - b);
                    dp[i].insert(a * b);
                    if (b != 0) dp[i].insert(a / b);
                }
            }
        }

        // 목표 값이 있는지 확인
        if (dp[i].count(number)) return i;
    }

    return -1;  // 8번보다 크면 -1 반환
}

File: 프로그래머스/3/42895. N으로 표현/N으로 표현.py
def solution(N, number):
    # dp[i]: set of numbers expressible with i uses of N
    dp = [set() for _ in range(9)]
    # Pre-fill concatenations: N, NN, NNN, ...
    for i in range(1, 9):
        dp[i].add(int(str(N) * i))

    # Build DP
    for i in range(1, 9):
        # Check if target achievable by concatenation alone
        if number in dp[i]:
            return i
        # Combine smaller counts
        for j in range(1, i):
            for op1 in dp[j]:
                for op2 in dp[i - j]:
                    dp[i].add(op1 + op2)
                    dp[i].add(op1 - op2)
                    dp[i].add(op2 - op1)
                    dp[i].add(op1 * op2)
                    if op2 != 0:
                        dp[i].add(op1 // op2)
                    if op1 != 0:
                        dp[i].add(op2 // op1)
        # After combining, check if target achievable
        if number in dp[i]:
            return i
    return -1